File: scripts/convert_webp.py
import os
import sys
import argparse
from pathlib import Path
from PIL import Image

if sys.platform == 'win32':
    try:
        sys.stdout.reconfigure(encoding='utf-8')
        sys.stderr.reconfigure(encoding='utf-8')
    except Exception:
        pass

def format_bytes(size):
    for unit in ['B', 'KB', 'MB', 'GB']:
        if size < 1024.0:
            return f"{size:.2f} {unit}"
        size /= 1024.0
    return f"{size:.2f} TB"

def convert_png_to_webp(target_path, quality=85, lossless=False, delete_original=False):
    target = Path(target_path)
    if not target.exists():
        print(f"❌ Caminho não encontrado: {target}")
        return

    if target.is_file():
        files = [target] if target.suffix.lower() == '.png' else []
    else:
        files = list(target.rglob('*.png'))

    if not files:
        print("⚠️  Nenhum arquivo .png encontrado.")
        return

    print(f"\n🖼️  Conversor PNG -> WebP (Python / Pillow)")
    print(f"📂 Alvo: {target.resolve()}")
    print(f"⚙️  Qualidade: {'Lossless' if lossless else quality} | Deletar originais: {'Sim' if delete_original else 'Não'}\n")
    print(f"Encontrados {len(files)} arquivo(s) PNG para converter...\n")

    total_original = 0
    total_converted = 0
    success = 0
    errors = 0

    for file_path in files:
        try:
            original_size = file_path.stat().st_size
            webp_path = file_path.with_suffix('.webp')

            with Image.open(file_path) as img:
                img.save(
                    webp_path,
                    format="WEBP",
                    quality=quality,
                    lossless=lossless,
                    method=6
                )

            new_size = webp_path.stat().st_size
            saved = original_size - new_size
            pct = (saved / original_size) * 100 if original_size > 0 else 0

            total_original += original_size
            total_converted += new_size
            success += 1

            sign = '-' if saved >= 0 else '+'
            print(f"  ✓ {file_path.name}")
            print(f"    {format_bytes(original_size)} -> {format_bytes(new_size)} ({sign}{abs(pct):.1f}%)\n")

            if delete_original:
                file_path.unlink()

        except Exception as e:
            errors += 1
            print(f"  ❌ Erro ao converter {file_path.name}: {e}\n")

    total_saved = total_original - total_converted
    total_pct = (total_saved / total_original * 100) if total_original > 0 else 0

    print("----------------------------------------------------")
    print(f"✅ Concluído: {success} convertidos{f', {errors} falhas' if errors > 0 else ''}.")
    print(f"📊 Tamanho total original: {format_bytes(total_original)}")
    print(f"📊 Tamanho total em WebP:  {format_bytes(total_converted)}")
    print(f"🎉 Economia total:        {format_bytes(total_saved)} ({total_pct:.1f}% de redução)")
    print("----------------------------------------------------\n")

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Converte arquivos PNG para WebP.")
    parser.add_argument("path", nargs="?", default=".", help="Caminho do arquivo ou diretório (padrão: pasta atual)")
    parser.add_argument("-q", "--quality", type=int, default=85, help="Qualidade WebP de 1 a 100 (padrão: 85)")
    parser.add_argument("--lossless", action="store_true", help="Conversão sem perdas (lossless)")
    parser.add_argument("--delete-original", "--delete", action="store_true", help="Exclui os arquivos .png originais após conversão")

    args = parser.parse_args()
    convert_png_to_webp(
        target_path=args.path,
        quality=args.quality,
        lossless=args.lossless,
        delete_original=args.delete_original
    )

File: scripts/convert-webp.js
import sharp from 'sharp';
import fs from 'fs/promises';
import { existsSync, statSync } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Formata bytes para formato legível (KB, MB)
function formatBytes(bytes) {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
}

// Parse argumentos simples da linha de comando
function parseArgs() {
  const args = process.argv.slice(2);
  const options = {
    targetPath: null,
    outDir: null,
    quality: 85,
    lossless: false,
    deleteOriginal: false,
    effort: 6,
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--lossless') {
      options.lossless = true;
    } else if (arg === '--delete-original' || arg === '--delete') {
      options.deleteOriginal = true;
    } else if (arg.startsWith('--quality=')) {
      options.quality = parseInt(arg.split('=')[1], 10);
    } else if (arg === '-q' || arg === '--quality') {
      options.quality = parseInt(args[++i], 10);
    } else if (arg.startsWith('--effort=')) {
      options.effort = parseInt(arg.split('=')[1], 10);
    } else if (arg.startsWith('--output=') || arg.startsWith('--out=')) {
      options.outDir = arg.split('=')[1];
    } else if (arg === '-o' || arg === '--output' || arg === '--outDir') {
      options.outDir = args[++i];
    } else if (!arg.startsWith('-') && !options.targetPath) {
      options.targetPath = arg;
    }
  }

  // Se nenhum caminho for passado, sugere o diretório padrão public/icons ou public
  if (!options.targetPath) {
    const defaultPublicIcons = path.resolve(__dirname, '../public/icons');
    const defaultPublic = path.resolve(__dirname, '../public');

    if (existsSync(defaultPublicIcons)) {
      options.targetPath = defaultPublicIcons;
    } else if (existsSync(defaultPublic)) {
      options.targetPath = defaultPublic;
    } else {
      options.targetPath = process.cwd();
    }
  } else {
    options.targetPath = path.resolve(process.cwd(), options.targetPath);
  }

  if (options.outDir) {
    options.outDir = path.resolve(process.cwd(), options.outDir);
  }

  return options;
}

// Coleta todos os arquivos .png recursivamente
async function getPngFiles(dirOrFile) {
  if (!existsSync(dirOrFile)) {
    throw new Error(`Caminho não encontrado: ${dirOrFile}`);
  }

  const stat = statSync(dirOrFile);
  if (stat.isFile()) {
    return dirOrFile.toLowerCase().endsWith('.png') ? [dirOrFile] : [];
  }

  const files = [];
  async function scan(dir) {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await scan(fullPath);
      } else if (entry.isFile() && entry.name.toLowerCase().endsWith('.png')) {
        files.push(fullPath);
      }
    }
  }

  await scan(dirOrFile);
  return files;
}

async function convertFile(pngPath, options) {
  const parsed = path.parse(pngPath);
  let webpPath;

  if (options.outDir) {
    const isTargetFile = statSync(options.targetPath).isFile();
    let destFolder = options.outDir;
    if (!isTargetFile) {
      const relDir = path.relative(options.targetPath, parsed.dir);
      destFolder = relDir ? path.join(options.outDir, relDir) : options.outDir;
    }
    if (!existsSync(destFolder)) {
      await fs.mkdir(destFolder, { recursive: true });
    }
    webpPath = path.join(destFolder, `${parsed.name}.webp`);
  } else {
    webpPath = path.join(parsed.dir, `${parsed.name}.webp`);
  }

  const initialStat = await fs.stat(pngPath);
  const originalSize = initialStat.size;

  const sharpInstance = sharp(pngPath);

  await sharpInstance
    .webp({
      quality: options.quality,
      lossless: options.lossless,
      effort: options.effort,
    })
    .toFile(webpPath);

  const newStat = await fs.stat(webpPath);
  const newSize = newStat.size;
  const savedBytes = originalSize - newSize;
  const reductionPct = ((savedBytes / originalSize) * 100).toFixed(1);

  if (options.deleteOriginal) {
    await fs.unlink(pngPath);
  }

  return {
    file: parsed.base,
    outputPath: webpPath,
    originalSize,
    newSize,
    savedBytes,
    reductionPct,
  };
}

async function main() {
  const options = parseArgs();

  console.log(`\n🖼️  Conversor PNG -> WebP`);
  console.log(`📂 Origem:  ${options.targetPath}`);
  if (options.outDir) {
    console.log(`📁 Destino: ${options.outDir}`);
  }
  console.log(`⚙️  Configurações: Qualidade: ${options.lossless ? 'Lossless' : options.quality} | Deletar originais: ${options.deleteOriginal ? 'Sim' : 'Não'}\n`);

  const files = await getPngFiles(options.targetPath);

  if (files.length === 0) {
    console.log('⚠️  Nenhum arquivo .png encontrado no caminho informado.');
    return;
  }

  console.log(`Encontrados ${files.length} arquivo(s) PNG para converter...\n`);

  let totalOriginal = 0;
  let totalConverted = 0;
  let successCount = 0;
  let errorCount = 0;

  for (const file of files) {
    try {
      const relativePath = path.relative(process.cwd(), file);
      const res = await convertFile(file, options);
      totalOriginal += res.originalSize;
      totalConverted += res.newSize;
      successCount++;

      const sign = res.savedBytes >= 0 ? '-' : '+';
      const absPct = Math.abs(res.reductionPct);
      const outputRel = path.relative(process.cwd(), res.outputPath);
      console.log(`  ✓ ${relativePath} -> ${outputRel}`);
      console.log(`    ${formatBytes(res.originalSize)} -> ${formatBytes(res.newSize)} (${sign}${absPct}%)\n`);
    } catch (err) {
      errorCount++;
      console.error(`  ❌ Erro ao converter ${file}:`, err.message);
    }
  }

  const totalSaved = totalOriginal - totalConverted;
  const totalReductionPct = totalOriginal > 0 ? ((totalSaved / totalOriginal) * 100).toFixed(1) : '0';

  console.log('----------------------------------------------------');
  console.log(`✅ Concluído: ${successCount} convertidos com sucesso${errorCount > 0 ? `, ${errorCount} falhas` : ''}.`);
  console.log(`📊 Tamanho total original: ${formatBytes(totalOriginal)}`);
  console.log(`📊 Tamanho total em WebP:  ${formatBytes(totalConverted)}`);
  console.log(`🎉 Economia total:        ${formatBytes(totalSaved)} (${totalReductionPct}% de redução)`);
  console.log('----------------------------------------------------\n');
}

main().catch(err => {
  console.error('Erro na execução:', err);
  process.exit(1);
});
